File: Difficulty: Medium/k-th Smallest in BST/kth-smallest-in-bst.cpp
/*
class Node {
  public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};
*/

class Solution {
  public:
    int kthSmallest(Node *root, int k) {
        // code here
        int count = 0;

        Node* curr = root;
        while (curr != nullptr) {
            if (curr->left == nullptr) {
                count++;
                if (count == k) return curr->data;
                curr = curr->right;
            } else {
                // Find the inorder predecessor of curr
                Node* prev = curr->left;
                while (prev->right != nullptr && prev->right != curr) {
                    prev = prev->right;
                }
    
                // Make curr the right child of its inorder predecessor
                if (prev->right == nullptr) {
                    prev->right = curr;
                    curr = curr->left;
                } else {
                    count++;
                    if (count == k) return curr->data;
    
                    // Revert the changes made in the tree structure
                    prev->right = nullptr;
                    curr = curr->right;
                }
            }
        }
        return -1;
    }
};

File: Difficulty: Medium/k-th Smallest in BST/kth-smallest-in-bst.java
/*
class Node {
    int data;
    Node left, right;

    public Node(int d)
    {
        data = d;
        left = right = null;
    }
}
*/

class Solution {
    int count=0;
    int kth=-1;
    public void inorder(Node root,int k){
        if(root==null||kth!=-1)return;
        inorder(root.left,k);
        count++;
        if(count==k){
            kth=root.data;
            return;
        }
        inorder(root.right,k);
    }
    public int kthSmallest(Node root, int k) {
        inorder(root,k);
        return kth;
    }
}